gframe/client_card.cpp

```diff
@@ -126,23 +126,90 @@ void ClientCard::ClearTarget() {
 	ownerTarget.clear();
 }
 bool ClientCard::client_card_sort(ClientCard* c1, ClientCard* c2) {
-	uint8_t cp1 = c1->overlayTarget ? c1->overlayTarget->controler : c1->controler;
-	uint8_t cp2 = c2->overlayTarget ? c2->overlayTarget->controler : c2->controler;
-	if(cp1 != cp2)
-		return cp1 < cp2;
-	if(c1->location != c2->location)
-		return c1->location < c2->location;
-	if(c1->location & LOCATION_OVERLAY) {
-		if(c1->overlayTarget != c2->overlayTarget)
-			return c1->overlayTarget->sequence < c2->overlayTarget->sequence;
+	// attached cards are sorted alongside the thing they are attached to
+	ClientCard* e1 = (c1->overlayTarget != nullptr) ? c1->overlayTarget : c1;
+	ClientCard* e2 = (c2->overlayTarget != nullptr) ? c2->overlayTarget : c2;
+	// if they are attached to the same thing, shortcut:
+	if (e1 == e2) {
+		if ((c1->overlayTarget != nullptr) != (c2->overlayTarget != nullptr)) {
+			// if only one is attached, the non-attached card comes first
+			return c1->overlayTarget == nullptr;
+		}
+		// if both are attached, order by sequence
 		return c1->sequence < c2->sequence;
 	}
-	if((c1->location & (LOCATION_DECK | LOCATION_GRAVE | LOCATION_REMOVED | LOCATION_EXTRA)) == 0)
-		return c1->sequence < c2->sequence;
-	for(const auto& chain : mainGame->dField.chains) {
-		if(c1 == chain.chain_card || chain.target.find(c1) != chain.target.end())
-			return true;
+	// player cards go before opponent cards
```

Comment:
upon review, i'm unsure whether `controler = 0` is always the player...

if not, is there a way to figure out who the player is in this context?